client/my-sites/upgrades/checkout-thank-you/index.jsx

```diff
@@ -139,60 +126,71 @@ var CheckoutThankYou = React.createClass( {
 		return getPurchases( this.props ).some( isJetpackPlan );
 	},
 
-	productRelatedMessages: function() {
-		var selectedSite = this.props.selectedSite,
-			purchases,
-			componentClass,
-			domain;
+	/**
+	 * Retrieves the component (and any corresponding data) that should be displayed according to the type of purchase
+	 * just performed by the user.
+	 *
+	 * @returns {*[]} an array of varying size with the component instance, then an optional purchase object possibly followed by a domain name
+	 */
+	getComponentAndPrimaryPurchaseAndDomain: function() {
+		if ( this.isDataLoaded() ) {
+			const purchases = getPurchases( this.props );
 
-		if ( ! this.isDataLoaded() ) {
-			return (
-				<div>
-					<PurchaseDetail isPlaceholder />
-					<PurchaseDetail isPlaceholder />
-					<PurchaseDetail isPlaceholder />
-				</div>
-			);
-		}
+			const findPurchaseAndDomain = ( purchases, predicate ) => {
+				const purchase = find( purchases, predicate );
 
-		if ( this.props.receiptId ) {
-			purchases = getPurchases( this.props );
+				return [ purchase, purchase.meta ];
+			};
 
 			if ( purchases.some( isJetpackPremium ) ) {
-				componentClass = JetpackPremiumPlanDetails;
+				return [ JetpackPremiumPlanDetails, find( purchases, isJetpackPremium ) ];
```

Comment:
The problem with returning an object is that this function will be more verbose because you'll have to add properties for each `return` statement. I agree it's cleaner though.